lib/stripe/credit_card_cloner.rb

```diff
@@ -1,38 +1,115 @@
 # frozen_string_literal: true
 
-# Here we clone
-#   - a card (card_*) or payment_method (pm_*) stored (in a customer) in a platform account
-# into
+# Here we clone (or find a clone of)
+#   - a card (card_*) or payment_method (pm_*) stored (in a customer) in a platform account into
 #   - a payment method (pm_*) (in a new customer) in a connected account
 #
 # This is required when using the Stripe Payment Intents API:
 #   - the customer and payment methods are stored in the platform account
 #       so that they can be re-used across multiple sellers
-#   - when a card needs to be charged, we need to create it in the seller's stripe account
+#   - when a card needs to be charged, we need to clone (or find the clone)
+#       in the seller's stripe account
 #
-# We are doing this process every time the card is charged:
-#   - this means that, if the customer uses the same card on the same seller multiple times,
-#       the card will be created multiple times on the seller's account
-#   - to avoid this, we would have to store the IDs of every card on each seller's stripe account
-#       in our database (this way we only have to store the platform account ID)
+# To avoid creating a new clone of the card/customer each time the card is charged or
+# authorized (e.g. for SCA), we attach metadata { clone: true } to the card the first time we
+# clone it and look for a card with the same fingerprint (hash of the card number) and
+# that metadata key to avoid cloning it multiple times.
+
 module Stripe
   class CreditCardCloner
+    def find_or_clone(card, connected_account_id)
+      if card.user && cloned_card = find_cloned_card(card, connected_account_id)
+        cloned_card
+      else
+        clone(card, connected_account_id)
+      end
+    end
+
+    def destroy_clones(card)
```

Comment:
I would rather move that to a separate service. This method has nothing to do with `#find_or_clone` and as a result, this class has (or will have) a bunch of private methods related to either `#find_or_clone` or `#destroy_clones`. Moreover, this may grow with more unrelated public methods over time looking more like a module with no state than a highly-cohesive class.

Reply:
IMO it's already sort of a code smell the fact that the top-level doc only refers to `#find_or_clone`

Reply:
I moved things to their own classes: 4c25edd91c395a22d592ad6ae7e0506e2879ca13, 8c747e4812195c9a75e34c7fdc248122329af9f4
And cleaned things up (a lot!) with ivars: b1dbb810536756bb1f584be443496f16ba59bc88